dilated-file-transfer.md

```diff
@@ -0,0 +1,476 @@
+Peers MUST only send messages for features in the other side's list.
+Only one feature exists currently: `"core0"` (and it MUST be supported).
+Peers MUST tolerate the existence of unknown values in the `"features"` list.
+
+   (XXX might make the Python implementation randomly add an unknown one, 10% of the time?)
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes or capabilities or evolving existing ones.
+
+The `"permission"` key specifies how to proceed with offers.
+Using mode `ask` tells the sender to pause after transmitting the first metadata message and await an answer from the peer before streaming data.
+Using mode `yes` means the peer will accept all incoming transfers so the sender should not pause after the metadata (and instead immediately start sending data messages).
+This cuts down latency for "one-way" transfers (see "Discussion")
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack` (rejected idea: JSON, because it lacks integers and binary types).
+
+Control-channel message formats are described using Python (and Haskell) pseudo-code to illustrate the exact data types involved.
+
+All control-channel messages contain a "kind" field describing the type of message.
+
+**Rejected idea**: Version message, because we already do version negotiation via mailbox features.
+
+**Rejected idea**: Offer/Answer messages via the control channel: we need to open a subchannel anyway; and the subchannel-IDs are not intended to be part of the public Dilation API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: "text"
+```
+
+All control-channel messages MUST be msgpack-encoded and include at least a `"kind"` field.
+Future extensions to the protocol may add additional control-channel messages (thus, any unknown control-channel messages MUST be ignored).
+To be clear, an unknown control-channel `"kind"` MUST NOT be treated as a protocol error.
+
+
+### Making an Offer
+
+Either side MAY send any number of Offer messages at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any offers.
+
+To make an offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+That is, a subchannel transmits a series of complete (framed) messages (up to ~4GiB in size).
+
+For this protocol, each record on the subchannel uses the first byte to indicate the kind of message; the remaining bytes in any message are kind-dependant.
```

Comment:
Perhaps this needs some more concrete numbers / testing before finalizing.

The reasoning is that you have to pass _all_ the bytes/memory through the msgpack codec then. With the "first byte" thing, you can efficiently (and predictably) chunk your payload data at 65KiB - 1 bytes on both encode and decode.

So when pulling data off, you'll bring "the message" into memory. Now you can simply look at the first byte, then write out the rest. With msgpack you'll end up allocating another 65KiB buffer for every message. (We could perhaps specify FlatBuffers here instead? But that has less widespread support and the zero-copy aspect is overkill for most of the messages ... )

Anyway, that's the thinking. I have not tested this, but I believe it would mean approximately twice as much memory pressure. (Perhaps actual msgpack libraries are more efficient than the api implies?)


I suppose -- similar to your reasoning above about the "mode" -- this could be left to a future extension if it is found to be substantially more efficient to do so. Overall, it seems unlikely that memory-allocator pressure will dominate runtime (although phone users might prefer less overhead). Again, completely untested :)

Reply:
Okay, yeah _at least_ that simplification makes sense (e.g. "just data" versus "msgpack-encoded data" in which case it must have a "kind" field).

I'll at least do _some_ tests and see if there's some obvious, large advantage. This could also be a "future enhancement" / exercise .